htmresearch/regions/ExtendedTMRegion.py

```diff
@@ -0,0 +1,449 @@
+                      "with internal cells within this temporal memory",
+          accessMode="ReadWrite",
+          dataType="UInt32",
+          count=1,
+          defaultValue=1,
+          constraints="bool"),
+        learnOnOneCell=dict(
+          description="If True, the winner cell for each column will be"
+                      " fixed between resets.",
+          accessMode="ReadWrite",
+          dataType="UInt32",
+          count=1,
+          constraints="bool"),
+        defaultOutputType=dict(
+          description="Controls what type of cell output is placed into"
+                      " the default output 'feedForwardOutput'",
+          accessMode="ReadWrite",
+          dataType="Byte",
+          count=0,
+          constraints="enum: active,predictive,predictedActiveCells",
+          defaultValue="active"),
+      ),
+      commands=dict(
+        reset=dict(description="Explicitly reset TM states now."),
+      )
+    )
+
+    return spec
+
+
+  def __init__(self,
+               columnCount=2048,
+               cellsPerColumn=32,
+               activationThreshold=13,
+               initialPermanence=0.21,
+               connectedPermanence=0.50,
+               minThreshold=10,
+               maxNewSynapseCount=20,
+               permanenceIncrement=0.10,
+               permanenceDecrement=0.10,
+               predictedSegmentDecrement=0.0,
+               seed=42,
+               learnOnOneCell=1,
+               formInternalConnections = 1,
+               defaultOutputType = "active",
+               **kwargs):
+    # Defaults for all other parameters
+
+    self.columnCount = columnCount
+    self.cellsPerColumn = cellsPerColumn
+    self.activationThreshold = activationThreshold
+    self.initialPermanence = initialPermanence
+    self.connectedPermanence = connectedPermanence
+    self.minThreshold = minThreshold
+    self.maxNewSynapseCount = maxNewSynapseCount
+    self.permanenceIncrement = permanenceIncrement
+    self.permanenceDecrement = permanenceDecrement
+    self.predictedSegmentDecrement = predictedSegmentDecrement
+    self.seed = seed
+    self.learnOnOneCell = bool(learnOnOneCell)
+    self.learningMode = True
+    self.inferenceMode = True
+    self.formInternalConnections = bool(formInternalConnections)
+    self.defaultOutputType = defaultOutputType
+
+    PyRegion.__init__(self, **kwargs)
+
+    # TM instance
+    self._tm = None
+
+
+  def initialize(self, inputs, outputs):
+    """
+    Initialize the self._tm if not already initialized. We need to figure out
+    the constructor parameters for each class, and send it to that constructor.
+    """
+    if self._tm is None:
+      # Create dict of arguments we will pass to the temporal memory class
+      args = copy.deepcopy(self.__dict__)
+      args["columnDimensions"] = (self.columnCount,)
+      args.pop("columnCount")
+
+      # Create the TM instance
+      self._tm = createModel("general", **args)
```

Comment:
Oh I wasn't familiar with this. And @natoromano just ran into a problem in SDRClassifierRegion where it was clobbering the deserialized instance in `initialize` so I was trying to avoid that issue. I'll think about how to make it clear to developers how to do this properly since right now it really isn't obvious example how to handle `__init__`/`initialize`.